File: Engine/src/main/java/org/etieskrill/engine/graphics/texture/font/BitmapFont.java
package org.etieskrill.engine.graphics.texture.font;

import org.etieskrill.engine.graphics.texture.Texture2D;

import java.util.Map;

public class BitmapFont implements Font {
    
    private final Map<Character, Glyph> glyphs;
    
    private final int lineHeight;
    private final int minLineHeight;
    
    private final String family;
    private final String style;
    
    public BitmapFont(Map<Character, Glyph> glyphs, int lineHeight, int minLineHeight, String family, String style) {
        this.glyphs = glyphs;
        this.lineHeight = lineHeight;
        this.minLineHeight = minLineHeight;
        this.family = family;
        this.style = style;
    }
    
    @Override
    public Glyph getGlyph(char c) {
        return glyphs.get(c);
    }
    
    @Override
    public Glyph[] getGlyphs(String s) {
        return s.codePoints().mapToObj(c -> getGlyph((char) c)).toArray(Glyph[]::new);
    }
    
    @Override
    public int getLineHeight() {
        return lineHeight;
    }
    
    @Override
    public int getMinLineHeight() {
        return minLineHeight;
    }
    
    public String getFamily() {
        return family;
    }
    
    public String getStyle() {
        return style;
    }
    
    @Override
    public void dispose() {
        glyphs.values().stream().map(Glyph::getTexture).forEach(Texture2D::dispose);
    }
    
}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/text/BitmapFont.kt
package org.etieskrill.engine.graphics.text

import org.etieskrill.engine.graphics.texture.ArrayTexture2D
import org.joml.Vector2ic

class BitmapFont(
    private val glyphs: Map<Char, Glyph>,

    override val lineHeight: Int,
    override val minLineHeight: Int,
    override val pixelSize: Vector2ic,

    val family: String,
    val style: String,

    val textures: ArrayTexture2D
) : Font {

    override fun getGlyph(c: Char): Glyph = glyphs[c] ?: glyphs[0.toChar()]!!

    override fun getGlyphs(s: String): Array<Glyph> = s.toCharArray().map { getGlyph(it) }.toTypedArray()

}
